=====================================================================
# <font size = 5 color='Black'><u>Employee Wellness Prediction using Logistic Regression<u> 
=====================================================================

## Abstract

Take care of your employees: A drastic thing happened and ABCXYZ123 Technical Solutions have lost one of their important employees. The company is now very concerned about the health of their employees and would want you to find that set of employees who are in need or may be in need of treatment, taking into account multiple attributes that are already stored in the database. So buckle up the wellness of your employees is in your hand.


## Problem Statement

The objective is to predict values “treatment” attribute from the given features of the Test data.

## Data Description

• Timestamp

• Age

• Gender

• Country

• state -> If you live in the United States, which state or territory do you live in?

• self_employed -> Are you self-employed?

• family_history -> Do you have a family history of mental illness?

• work_interfere -> If you have a mental health condition, do you feel that it interferes with your work?

• no_employees -> How many employees does your company or organization have?

• remote_work -> Do you work remotely (outside of an office) at least 50% of the time?

• tech_company -> Is your employer primarily a tech company/organization?

• benefits -> Does your employer provide mental health benefits?

• care_options -> Do you know the options for mental health care your employer provides?

• wellness_program -> Has your employer ever discussed mental health as part of an employee wellness program?

• seek_help -> Does your employer provide resources to learn more about mental health issues and how to seek help?

• anonymity -> Is your anonymity protected if you choose to take advantage of mental health or substance abuse treatment resources?

• leave -> How easy is it for you to take medical leave for a mental health condition?

• mental_health_consequence -> Do you think that discussing a mental health issue with your employer would have negative consequences?

• phys_health_consequence -> Do you think that discussing a physical health issue with your employer would have negative consequences?

• coworkers -> Would you be willing to discuss a mental health issue with your coworkers?

• supervisor -> Would you be willing to discuss a mental health issue with your direct supervisor(s)?

• mental_health_interview -> Would you bring up a mental health issue with a potential employer in an interview?

• phys_health_interview -> Would you bring up a physical health issue with a potential employer in an interview?

• mental_vs_physical -> Do you feel that your employer takes mental health as seriously as physical health?

• obs_consequence -> Have you heard of or observed negative consequences for coworkers with mental health conditions in your workplace?

• comments -> Any additional notes or comments.

• treatment -> Does he or she really needs treatment.

=====================================================================

### Importing Necessary Libraries

In [29]:
#to hide warnings
import warnings
warnings.filterwarnings("ignore")

In [30]:
#for working with arrays
import numpy as np
#for reading data and other manipulations
import pandas as pd
#for visulaization
import matplotlib.pyplot as plt
import seaborn as sns
#for encoding categorical variables
from sklearn.preprocessing import LabelEncoder
#for scaling the data
from sklearn.preprocessing import StandardScaler
#for splitting the data into train set ans test set
from sklearn.model_selection import train_test_split
#for performing Linear regression and other regularization techniques
from sklearn.linear_model import LogisticRegression
#for model evaluation
from sklearn.metrics import confusion_matrix,classification_report,accuracy_score
#for performing Stochastic Gradient Descent Classification
from sklearn.linear_model import SGDClassifier
#for feature selection
from sklearn.feature_selection import RFE
#for kfold cross validation
from sklearn.model_selection import KFold
from sklearn.model_selection import cross_val_score

### Reading data 

In [31]:
#Creating a dataframe for train data
data = pd.read_csv(r"C:/Users/odhonde/OneDrive - DXC Production/Documents/Imeticas Learning/Hackthon_13_05_2023/employee_train.csv",header = 0,delimiter = ',')
#Creating a dataframe for test data
test_data = pd.read_csv(r"C:/Users/odhonde/OneDrive - DXC Production/Documents/Imeticas Learning/Hackthon_13_05_2023/employee_test.csv",header = 0,delimiter = ',')

In [32]:
#to display all the columns and intended rows on the screen
pd.set_option('display.max_columns',None)
#to check the sample data
data.head()

S.No        Timestamp  Age  Gender         Country state self_employed  \
0     1  8/27/2014 11:29   37  Female   United States    IL           NaN   
1     2  8/27/2014 11:29   44       M   United States    IN           NaN   
2     3  8/27/2014 11:29   32    Male          Canada   NaN           NaN   
3     4  8/27/2014 11:29   31    Male  United Kingdom   NaN           NaN   
4     5  8/27/2014 11:30   31    Male   United States    TX           NaN   

  family_history treatment work_interfere    no_employees remote_work  \
0             No       Yes          Often          25-Jun          No   
1             No        No         Rarely  More than 1000          No   
2             No        No         Rarely          25-Jun          No   
3            Yes       Yes          Often          26-100          No   
4             No        No          Never         100-500         Yes   

  tech_company    benefits care_options wellness_program   seek_help  \
0          Yes         Yes     Not sure               No         Yes   
1           No  Don't know           No       Don't know  Don't know   
2          Yes          No           No               No          No   
3          Yes          No          Yes               No          No   
4          Yes         Yes           No       Don't know  Don't know   

    anonymity               leave mental_health_consequence  \
0         Yes       Somewhat easy                        No   
1  Don't know          Don't know                     Maybe   
2  Don't know  Somewhat difficult                        No   
3          No  Somewhat difficult                       Yes   
4  Don't know          Don't know                        No   

  phys_health_consequence     coworkers supervisor mental_health_interview  \
0                      No  Some of them        Yes                      No   
1                      No            No         No                      No   
2                      No           Yes        Yes                     Yes   
3                     Yes  Some of them         No                   Maybe   
4                      No  Some of them        Yes                     Yes   

  phys_health_interview mental_vs_physical obs_consequence comments  
0                 Maybe                Yes              No      NaN  
1                    No         Don't know              No      NaN  
2                   Yes                 No              No      NaN  
3                 Maybe                 No             Yes      NaN  
4                   Yes         Don't know              No      NaN

In [33]:
test_data.head()

S.No        Timestamp  Age  Gender         Country state self_employed  \
0     1  8/29/2014 11:32   39    Male  United Kingdom   NaN           Yes   
1     2  8/29/2014 11:32   26  female   United States    WA            No   
2     3  8/29/2014 11:33   23  Female   United States    IL            No   
3     4  8/29/2014 11:34   35    Male     Switzerland   NaN            No   
4     5  8/29/2014 11:36   36    Male   United States    FL            No   

  family_history work_interfere    no_employees remote_work tech_company  \
0            Yes      Sometimes           5-Jan         Yes          Yes   
1            Yes      Sometimes  More than 1000          No          Yes   
2            Yes      Sometimes          26-100          No           No   
3            Yes          Often  More than 1000          No          Yes   
4             No          Never           5-Jan         Yes          Yes   

     benefits care_options wellness_program   seek_help   anonymity  \
0          No          Yes              Yes         Yes         Yes   
1         Yes          Yes               No          No  Don't know   
2          No           No               No          No  Don't know   
3  Don't know     Not sure               No          No         Yes   
4  Don't know     Not sure       Don't know  Don't know  Don't know   

                leave mental_health_consequence phys_health_consequence  \
0  Somewhat difficult                        No                      No   
1          Don't know                        No                      No   
2  Somewhat difficult                       Yes                      No   
3           Very easy                        No                      No   
4           Very easy                        No                      No   

      coworkers    supervisor mental_health_interview phys_health_interview  \
0           Yes           Yes                      No                 Maybe   
1  Some of them           Yes                      No                 Maybe   
2            No  Some of them                      No                 Maybe   
3  Some of them  Some of them                      No                 Maybe   
4  Some of them  Some of them                      No                    No   

  mental_vs_physical obs_consequence  \
0                Yes             Yes   
1                 No             Yes   
2                 No              No   
3                 No              No   
4         Don't know              No   

                                            comments  
0  These result may be a tad confusing so a summa...  
1  I should note one of the places my employer fa...  
2                                                NaN  
3                                                NaN  
4                                                NaN

### Understanding the data

In [34]:
data.shape

(1048, 28)

Data has 1048 observation and 28 columns out of which 1 column has Dependent variable and other 27 are Independent var

In [35]:
data.columns

Index(['S.No', 'Timestamp', 'Age', 'Gender', 'Country', 'state',
       'self_employed', 'family_history', 'treatment', 'work_interfere',
       'no_employees', 'remote_work', 'tech_company', 'benefits',
       'care_options', 'wellness_program', 'seek_help', 'anonymity', 'leave',
       'mental_health_consequence', 'phys_health_consequence', 'coworkers',
       'supervisor', 'mental_health_interview', 'phys_health_interview',
       'mental_vs_physical', 'obs_consequence', 'comments'],
      dtype='object')

In [36]:
data.dtypes

S.No                          int64
Timestamp                    object
Age                           int64
Gender                       object
Country                      object
state                        object
self_employed                object
family_history               object
treatment                    object
work_interfere               object
no_employees                 object
remote_work                  object
tech_company                 object
benefits                     object
care_options                 object
wellness_program             object
seek_help                    object
anonymity                    object
leave                        object
mental_health_consequence    object
phys_health_consequence      object
coworkers                    object
supervisor                   object
mental_health_interview      object
phys_health_interview        object
mental_vs_physical           object
obs_consequence              object
comments                    

we can see that most of the variables are categorical variables

In [37]:
#to get insights from data by looking at their mean,std etc.,
data.describe(include = 'all')

S.No        Timestamp           Age Gender        Country  \
count   1048.000000             1048  1.048000e+03   1048           1048   
unique          NaN              679           NaN     45             42   
top             NaN  8/27/2014 15:24           NaN   Male  United States   
freq            NaN                8           NaN    518            644   
mean     524.500000              NaN  9.541988e+07    NaN            NaN   
std      302.675844              NaN  3.089010e+09    NaN            NaN   
min        1.000000              NaN -1.726000e+03    NaN            NaN   
25%      262.750000              NaN  2.700000e+01    NaN            NaN   
50%      524.500000              NaN  3.100000e+01    NaN            NaN   
75%      786.250000              NaN  3.600000e+01    NaN            NaN   
max     1048.000000              NaN  1.000000e+11    NaN            NaN   

       state self_employed family_history treatment work_interfere  \
count    636          1030           1048      1048            812   
unique    45             2              2         2              4   
top       CA            No             No        No      Sometimes   
freq     123           906            643       536            386   
mean     NaN           NaN            NaN       NaN            NaN   
std      NaN           NaN            NaN       NaN            NaN   
min      NaN           NaN            NaN       NaN            NaN   
25%      NaN           NaN            NaN       NaN            NaN   
50%      NaN           NaN            NaN       NaN            NaN   
75%      NaN           NaN            NaN       NaN            NaN   
max      NaN           NaN            NaN       NaN            NaN   

       no_employees remote_work tech_company benefits care_options  \
count          1048        1048         1048     1035         1048   
unique            6           2            2        3            3   
top          25-Jun          No          Yes      Yes           No   
freq            245         733          870      400          422   
mean            NaN         NaN          NaN      NaN          NaN   
std             NaN         NaN          NaN      NaN          NaN   
min             NaN         NaN          NaN      NaN          NaN   
25%             NaN         NaN          NaN      NaN          NaN   
50%             NaN         NaN          NaN      NaN          NaN   
75%             NaN         NaN          NaN      NaN          NaN   
max             NaN         NaN          NaN      NaN          NaN   

       wellness_program seek_help   anonymity       leave  \
count              1044      1048        1048        1044   
unique                3         3           3           5   
top                  No        No  Don't know  Don't know   
freq                692       534         685         462   
mean                NaN       NaN         NaN         NaN   
std                 NaN       NaN         NaN         NaN   
min                 NaN       NaN         NaN         NaN   
25%                 NaN       NaN         NaN         NaN   
50%                 NaN       NaN         NaN         NaN   
75%                 NaN       NaN         NaN         NaN   
max                 NaN       NaN         NaN         NaN   

       mental_health_consequence phys_health_consequence     coworkers  \
count                       1048                    1048          1048   
unique                         3                       3             3   
top                           No                      No  Some of them   
freq                         424                     773           651   
mean                         NaN                     NaN           NaN   
std                          NaN                     NaN           NaN   
min                          NaN                     NaN           NaN   
25%                          NaN                     NaN           NaN   
50%                  

In [38]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1048 entries, 0 to 1047
Data columns (total 28 columns):
 #   Column                     Non-Null Count  Dtype 
---  ------                     --------------  ----- 
 0   S.No                       1048 non-null   int64 
 1   Timestamp                  1048 non-null   object
 2   Age                        1048 non-null   int64 
 3   Gender                     1048 non-null   object
 4   Country                    1048 non-null   object
 5   state                      636 non-null    object
 6   self_employed              1030 non-null   object
 7   family_history             1048 non-null   object
 8   treatment                  1048 non-null   object
 9   work_interfere             812 non-null    object
 10  no_employees               1048 non-null   object
 11  remote_work                1048 non-null   object
 12  tech_company               1048 non-null   object
 13  benefits                   1035 non-null   object
 14  care_opt

As per the Non-Null Count, we can see there are a lot of missing values in some columns like state, work_interfere and comments

In [39]:
# to check for the unique values in all the variables to understand the data
#we do not want unique values for timestamp and comments because they are going to be enormous
for i in data.columns:
    if i != 'Timestamp' and i != 'comments':
        print("COLUMN: ", i.upper())
        print("")
        print(data[i].unique())
        print("")

COLUMN:  S.NO

[   1    2    3 ... 1046 1047 1048]

COLUMN:  AGE

[         37          44          32          31          33          35
          39          42          23          29          36          27
          46          41          34          30          40          38
          50          24          18          28          26          22
          19          25          45          21         -29          43
          56          60          54         329          55 99999999999
          48          20          57          58          47          62
          51          65          49       -1726           5          53
          61           8]

COLUMN:  GENDER

['Female' 'M' 'Male' 'male' 'female' 'm' 'Male-ish' 'maile' 'Trans-female'
 'Cis Female' 'F' 'something kinda male?' 'Cis Male' 'Woman' 'f' 'Mal'
 'Male (CIS)' 'queer/she/they' 'non-binary' 'Femake' 'woman' 'Make' 'Nah'
 'All' 'Enby' 'fluid' 'Genderqueer' 'Female ' 'Androgyne' 'Agender'
 'cis-female/femme

We can see that we need to encode all the categorical variables to discrete/numerical variables

And we also noticed that some values of age like 99999999999,329,-1726,-29 are totally inaccurate and we should handle them

### Data Preprocessing

# <font size = 3 color='Black'><b>1). Feature Selection</b>

We can drop the insignificant independent variableby using domain specific knowledge

In this data, Sno,Timestamp,comments,country,state does not have any significance with respect to predictiong wellness of employees

In [40]:
#we'll drop the data from the copy of dataframe to keep the original data intact
df = data.drop(['S.No', 'Timestamp','Country', 'state','comments'], axis =1)
#will apply same to the test data
df_test = test_data.drop(['S.No', 'Timestamp','Country', 'state','comments'], axis =1)

In [41]:
df.head()

Age  Gender self_employed family_history treatment work_interfere  \
0   37  Female           NaN             No       Yes          Often   
1   44       M           NaN             No        No         Rarely   
2   32    Male           NaN             No        No         Rarely   
3   31    Male           NaN            Yes       Yes          Often   
4   31    Male           NaN             No        No          Never   

     no_employees remote_work tech_company    benefits care_options  \
0          25-Jun          No          Yes         Yes     Not sure   
1  More than 1000          No           No  Don't know           No   
2          25-Jun          No          Yes          No           No   
3          26-100          No          Yes          No          Yes   
4         100-500         Yes          Yes         Yes           No   

  wellness_program   seek_help   anonymity               leave  \
0               No         Yes         Yes       Somewhat easy   
1       Don't know  Don't know  Don't know          Don't know   
2               No          No  Don't know  Somewhat difficult   
3               No          No          No  Somewhat difficult   
4       Don't know  Don't know  Don't know          Don't know   

  mental_health_consequence phys_health_consequence     coworkers supervisor  \
0                        No                      No  Some of them        Yes   
1                     Maybe                      No            No         No   
2                        No                      No           Yes        Yes   
3                       Yes                     Yes  Some of them         No   
4                        No                      No  Some of them        Yes   

  mental_health_interview phys_health_interview mental_vs_physical  \
0                      No                 Maybe                Yes   
1                      No                    No         Don't know   
2                     Yes                   Yes                 No   
3                   Maybe                 Maybe                 No   
4                     Yes                   Yes         Don't know   

  obs_consequence  
0              No  
1              No  
2              No  
3             Yes  
4              No

In [42]:
df_test.head()

Age  Gender self_employed family_history work_interfere    no_employees  \
0   39    Male           Yes            Yes      Sometimes           5-Jan   
1   26  female            No            Yes      Sometimes  More than 1000   
2   23  Female            No            Yes      Sometimes          26-100   
3   35    Male            No            Yes          Often  More than 1000   
4   36    Male            No             No          Never           5-Jan   

  remote_work tech_company    benefits care_options wellness_program  \
0         Yes          Yes          No          Yes              Yes   
1          No          Yes         Yes          Yes               No   
2          No           No          No           No               No   
3          No          Yes  Don't know     Not sure               No   
4         Yes          Yes  Don't know     Not sure       Don't know   

    seek_help   anonymity               leave mental_health_consequence  \
0         Yes         Yes  Somewhat difficult                        No   
1          No  Don't know          Don't know                        No   
2          No  Don't know  Somewhat difficult                       Yes   
3          No         Yes           Very easy                        No   
4  Don't know  Don't know           Very easy                        No   

  phys_health_consequence     coworkers    supervisor mental_health_interview  \
0                      No           Yes           Yes                      No   
1                      No  Some of them           Yes                      No   
2                      No            No  Some of them                      No   
3                      No  Some of them  Some of them                      No   
4                      No  Some of them  Some of them                      No   

  phys_health_interview mental_vs_physical obs_consequence  
0                 Maybe                Yes             Yes  
1                 Maybe                 No             Yes  
2                 Maybe                 No              No  
3                 Maybe                 No              No  
4                    No         Don't know              No

# <font size = 3 color='Black'><b>2). Handling Missing Values</b>

In [43]:
#Check for missing values count in whole dataset
df.isnull().sum()

Age                            0
Gender                         0
self_employed                 18
family_history                 0
treatment                      0
work_interfere               236
no_employees                   0
remote_work                    0
tech_company                   0
benefits                      13
care_options                   0
wellness_program               4
seek_help                      0
anonymity                      0
leave                          4
mental_health_consequence      0
phys_health_consequence        0
coworkers                      0
supervisor                     0
mental_health_interview        0
phys_health_interview          0
mental_vs_physical             0
obs_consequence                0
dtype: int64

We can see there are Nan values in self_emplyed and work_interfere

As these are categorical variables, we'replace Nan with mode

In [44]:
df_test.isnull().sum()

Age                           0
Gender                        0
self_employed                 0
family_history                0
work_interfere               28
no_employees                  0
remote_work                   0
tech_company                  0
benefits                      3
care_options                  0
wellness_program              2
seek_help                     0
anonymity                     0
leave                         5
mental_health_consequence     0
phys_health_consequence       0
coworkers                     0
supervisor                    0
mental_health_interview       0
phys_health_interview         0
mental_vs_physical            0
obs_consequence               0
dtype: int64

There are missing values in work_interfere in test data

In [45]:
#we can use use fillna to fill the missing values with mean of the variable
#inplace is kept as true to work om only the missing values rather than the whole data and thereby reducing kernel load
df['self_employed'].fillna(df['self_employed'].mode()[0], inplace = True)

In [46]:
#df['work_interfere'].fillna(df['work_interfere'].mode()[0], inplace = True)
#as work_interfere is multi class variable so we added a new class for NaN values

In [47]:
df['work_interfere'].fillna("Don't Know", inplace = True)

In [48]:
df.isnull().sum()

Age                           0
Gender                        0
self_employed                 0
family_history                0
treatment                     0
work_interfere                0
no_employees                  0
remote_work                   0
tech_company                  0
benefits                     13
care_options                  0
wellness_program              4
seek_help                     0
anonymity                     0
leave                         4
mental_health_consequence     0
phys_health_consequence       0
coworkers                     0
supervisor                    0
mental_health_interview       0
phys_health_interview         0
mental_vs_physical            0
obs_consequence               0
dtype: int64

In [49]:
df['benefits'].dtype

dtype('O')

In [50]:
df['benefits'].value_counts()

Yes           400
Don't know    329
No            306
Name: benefits, dtype: int64

In [51]:
df['benefits'].fillna(df['benefits'].mode()[0], inplace = True)

In [52]:
df['wellness_program'].value_counts()

No            692
Yes           193
Don't know    159
Name: wellness_program, dtype: int64

In [53]:
df['wellness_program'].fillna(df['wellness_program'].mode()[0], inplace = True)

In [54]:
df['leave'].value_counts()

Don't know            462
Somewhat easy         218
Very easy             176
Somewhat difficult    107
Very difficult         81
Name: leave, dtype: int64

In [55]:
df['leave'].fillna(df['leave'].mode()[0], inplace = True)

In [56]:
df.isnull().sum()

Age                          0
Gender                       0
self_employed                0
family_history               0
treatment                    0
work_interfere               0
no_employees                 0
remote_work                  0
tech_company                 0
benefits                     0
care_options                 0
wellness_program             0
seek_help                    0
anonymity                    0
leave                        0
mental_health_consequence    0
phys_health_consequence      0
coworkers                    0
supervisor                   0
mental_health_interview      0
phys_health_interview        0
mental_vs_physical           0
obs_consequence              0
dtype: int64

In [57]:
df_test['work_interfere'].fillna("Don't Know", inplace = True)

In [58]:
df_test.isnull().sum()

Age                          0
Gender                       0
self_employed                0
family_history               0
work_interfere               0
no_employees                 0
remote_work                  0
tech_company                 0
benefits                     3
care_options                 0
wellness_program             2
seek_help                    0
anonymity                    0
leave                        5
mental_health_consequence    0
phys_health_consequence      0
coworkers                    0
supervisor                   0
mental_health_interview      0
phys_health_interview        0
mental_vs_physical           0
obs_consequence              0
dtype: int64

In [59]:
df_test['benefits'].value_counts()

Yes           77
No            67
Don't know    63
Name: benefits, dtype: int64

In [60]:
df_test['benefits'].fillna(df_test['benefits'].mode()[0], inplace = True)

In [61]:
df_test['wellness_program'].value_counts()

No            150
Yes            35
Don't know     23
Name: wellness_program, dtype: int64

In [62]:
df_test['wellness_program'].fillna(df_test['wellness_program'].mode()[0], inplace = True)

In [63]:
df_test['leave'].value_counts()

Don't know            92
Somewhat easy         48
Very easy             30
Somewhat difficult    19
Very difficult        16
Name: leave, dtype: int64

In [64]:
df_test['leave'].fillna(df_test['leave'].mode()[0], inplace = True)

In [65]:
df_test.isnull().sum()

Age                          0
Gender                       0
self_employed                0
family_history               0
work_interfere               0
no_employees                 0
remote_work                  0
tech_company                 0
benefits                     0
care_options                 0
wellness_program             0
seek_help                    0
anonymity                    0
leave                        0
mental_health_consequence    0
phys_health_consequence      0
coworkers                    0
supervisor                   0
mental_health_interview      0
phys_health_interview        0
mental_vs_physical           0
obs_consequence              0
dtype: int64

# <font size = 3 color='Black'><b>Handling Age variable</b>

The possible age range for employees can be considered between 17 to 65 and therefore we must replace/delete the values above or below this range

In [66]:
#calculating mean for values between 17 to 65
mean_age = int(df['Age'][(df['Age']<=65) & (df['Age']>=17)].mean())
mean_age_test = int(df_test['Age'][(df_test['Age']<=65) & (df_test['Age']>=17)].mean())

In [67]:
#Replacing values greater than 65 or lower than 17 with mean
df['Age'] = df['Age'].apply(lambda x: mean_age if (x>65 or x<17) else int(x))
df_test['Age']= df_test['Age'].apply(lambda x: mean_age_test if(x>65 or x<17) else int(x))

# <font size = 3 color='Black'><b>3). Converting Categorical/Discrete Var to Continuous Var</b>

We need to convert all variables except Age to discrete/numerical variables


1.) Gender

We need to manually encode the values to female(0),male(1) and others(2) as there are too many class for single labels

'''['Female' 'M' 'Male' 'male' 'female' 'm' 'Male-ish' 'maile' 'Trans-female'
 'Cis Female' 'F' 'something kinda male?' 'Cis Male' 'Woman' 'f' 'Mal'
 'Male (CIS)' 'queer/she/they' 'non-binary' 'Femake' 'woman' 'Make' 'Nah'
 'All' 'Enby' 'fluid' 'Genderqueer' 'Female ' 'Androgyne' 'Agender'
 'cis-female/femme' 'Guy (-ish) ^_^' 'male leaning androgynous' 'Male '
 'Man' 'Trans woman' 'msle' 'Neuter' 'Female (trans)' 'queer'
 'Female (cis)' 'Mail' 'cis male' 'A little about you' 'Malr']'''

In [68]:
df['Gender'] = df['Gender'].replace(('Female','female','F','Woman','f','Femake','woman','Female ','Cis Female','cis-female/femme','Female (cis)'),0)
df['Gender'] = df['Gender'].replace(('M','Male','male','m','Male-ish','maile','Mal','Make','Male ','Man','msle','Mail','Malr','Cis Male','Male (CIS)','cis male'),1)
df['Gender'] = df['Gender'].replace(('Trans-female','something kinda male?','queer/she/they','non-binary',
                                 'Nah','All','Enby','fluid','Genderqueer','Androgyne','Agender','Guy (-ish) ^_^',
                                  'male leaning androgynous','Trans woman','Neuter','Female (trans)','queer',
                                  'A little about you'),2)

In [69]:
df['Gender'].unique()

array([0, 1, 2], dtype=int64)

In [70]:
# to check for the unique values of gender in test data
df_test['Gender'].unique()

array(['Male', 'female', 'Female', 'M', 'male', 'Male ', 'm', 'p', 'F',
       'Woman', 'femail', 'Cis Man',
       'ostensibly male, unsure what that really means', 'f'],
      dtype=object)

In [71]:
#Applying encoding to test data
df_test['Gender'] = df_test['Gender'].replace(('Female','female','F','Woman','f','femail'),0)
df_test['Gender'] = df_test['Gender'].replace(('M','Male','male','m','Male ','Cis Man'),1)
df_test['Gender'] = df_test['Gender'].replace(('p','ostensibly male, unsure what that really means'),2)

In [72]:
df_test['Gender'].unique()

array([1, 0, 2], dtype=int64)

2.) no_employees

Unique values: ['6-25' 'More than 1000' '26-100' '100-500' '1-5' '500-1000']

In [73]:
df['no_employees'].unique()

array(['25-Jun', 'More than 1000', '26-100', '100-500', '5-Jan',
       '500-1000'], dtype=object)

In [75]:
df['no_employees'] = df['no_employees'].replace('25-Jun', '6-25')

In [76]:
df['no_employees'] = df['no_employees'].replace('5-Jan', '1-5')

In [77]:
df['no_employees'].unique()

array(['6-25', 'More than 1000', '26-100', '100-500', '1-5', '500-1000'],
      dtype=object)

In [78]:
df['no_employees'] = df['no_employees'].map({'1-5':0,'6-25':1,'26-100':2,'100-500':3,'500-1000':4,'More than 1000':5})

In [79]:
df_test['no_employees'].unique()

array(['5-Jan', 'More than 1000', '26-100', '500-1000', '25-Jun',
       '100-500'], dtype=object)

In [80]:
df_test['no_employees'] = df_test['no_employees'].replace('25-Jun', '6-25')

In [81]:
df_test['no_employees'] = df_test['no_employees'].replace('5-Jan', '1-5')

In [82]:
df_test['no_employees'].unique()

array(['1-5', 'More than 1000', '26-100', '500-1000', '6-25', '100-500'],
      dtype=object)

In [83]:
df_test['no_employees'] = df_test['no_employees'].map({'1-5':0,'6-25':1,'26-100':2,'100-500':3,'500-1000':4,'More than 1000':5})

In [84]:
df_test.dtypes

Age                           int64
Gender                        int64
self_employed                object
family_history               object
work_interfere               object
no_employees                  int64
remote_work                  object
tech_company                 object
benefits                     object
care_options                 object
wellness_program             object
seek_help                    object
anonymity                    object
leave                        object
mental_health_consequence    object
phys_health_consequence      object
coworkers                    object
supervisor                   object
mental_health_interview      object
phys_health_interview        object
mental_vs_physical           object
obs_consequence              object
dtype: object

3.)All other categorical values

In [85]:
df['leave'].unique()

array(['Somewhat easy', "Don't know", 'Somewhat difficult',
       'Very difficult', 'Very easy'], dtype=object)

In [86]:
#Reducing redundant class to a single class for 'leaves' variable
df['leave'] = df['leave'].replace(['Somewhat easy', 'Very easy'],'easy')
df_test['leave'] = df_test['leave'].replace(['Somewhat easy', 'Very easy'],'easy')

In [87]:
encode = LabelEncoder()
for i in df.columns:
    if df[i].dtype == 'object':
            df[i] = encode.fit_transform(df[i])

In [88]:
#To check if all the data types have been converted to int from object
df.dtypes

Age                          int64
Gender                       int64
self_employed                int32
family_history               int32
treatment                    int32
work_interfere               int32
no_employees                 int64
remote_work                  int32
tech_company                 int32
benefits                     int32
care_options                 int32
wellness_program             int32
seek_help                    int32
anonymity                    int32
leave                        int32
mental_health_consequence    int32
phys_health_consequence      int32
coworkers                    int32
supervisor                   int32
mental_health_interview      int32
phys_health_interview        int32
mental_vs_physical           int32
obs_consequence              int32
dtype: object

In [89]:
df.head()

Age  Gender  self_employed  family_history  treatment  work_interfere  \
0   37       0              0               0          1               2   
1   44       1              0               0          0               3   
2   32       1              0               0          0               3   
3   31       1              0               1          1               2   
4   31       1              0               0          0               1   

   no_employees  remote_work  tech_company  benefits  care_options  \
0             1            0             1         2             1   
1             5            0             0         0             0   
2             1            0             1         1             0   
3             2            0             1         1             2   
4             3            1             1         2             0   

   wellness_program  seek_help  anonymity  leave  mental_health_consequence  \
0                 1          2          2      3                          1   
1                 0          0          0      0                          0   
2                 1          1          0      1                          1   
3                 1          1          1      1                          2   
4                 0          0          0      0                          1   

   phys_health_consequence  coworkers  supervisor  mental_health_interview  \
0                        1          1           2                        1   
1                        1          0           0                        1   
2                        1          2           2                        2   
3                        2          1           0                        0   
4                        1          1           2                        2   

   phys_health_interview  mental_vs_physical  obs_consequence  
0                      0                   2                0  
1                      1                   0                0  
2                      2                   1                0  
3                      0                   1                1  
4                      2                   0                0

In [90]:
df.tail()

Age  Gender  self_employed  family_history  treatment  work_interfere  \
1043   26       1              0               0          1               4   
1044   29       0              0               0          1               2   
1045   26       0              0               1          1               4   
1046   33       1              0               1          1               4   
1047   28       1              0               1          0               4   

      no_employees  remote_work  tech_company  benefits  care_options  \
1043             1            0             1         2             2   
1044             5            0             0         2             2   
1045             3            0             1         2             1   
1046             1            0             1         1             1   
1047             3            0             1         1             0   

      wellness_program  seek_help  anonymity  leave  \
1043                 1          1          2      0   
1044                 1          1          0      0   
1045                 1          1          2      0   
1046                 1          1          0      3   
1047                 1          1          0      3   

      mental_health_consequence  phys_health_consequence  coworkers  \
1043                          0                        1          1   
1044                          0                        1          1   
1045                          0                        1          1   
1046                          1                        1          1   
1047                          1                        1          1   

      supervisor  mental_health_interview  phys_health_interview  \
1043           0                        1                      0   
1044           0                        1                      2   
1045           0                        1                      0   
1046           2                        1                      1   
1047           1                        1                      1   

      mental_vs_physical  obs_consequence  
1043                   0                0  
1044                   1                1  
1045                   0                0  
1046                   0                0  
1047                   0                1

In [91]:
#encoding variables for data from test set
for i in df_test.columns:
    if df_test[i].dtype == 'object':
            df_test[i] = encode.fit_transform(df_test[i])

In [92]:
#To check if all the data types have been converted to int from object
df_test.dtypes

Age                          int64
Gender                       int64
self_employed                int32
family_history               int32
work_interfere               int32
no_employees                 int64
remote_work                  int32
tech_company                 int32
benefits                     int32
care_options                 int32
wellness_program             int32
seek_help                    int32
anonymity                    int32
leave                        int32
mental_health_consequence    int32
phys_health_consequence      int32
coworkers                    int32
supervisor                   int32
mental_health_interview      int32
phys_health_interview        int32
mental_vs_physical           int32
obs_consequence              int32
dtype: object

In [93]:
df_test.head()

Age  Gender  self_employed  family_history  work_interfere  no_employees  \
0   39       1              1               1               4             0   
1   26       0              0               1               4             5   
2   23       0              0               1               4             2   
3   35       1              0               1               2             5   
4   36       1              0               0               1             0   

   remote_work  tech_company  benefits  care_options  wellness_program  \
0            1             1         1             2                 2   
1            0             1         2             2                 1   
2            0             0         1             0                 1   
3            0             1         0             1                 1   
4            1             1         0             1                 0   

   seek_help  anonymity  leave  mental_health_consequence  \
0          2          2      1                          1   
1          1          0      0                          1   
2          1          0      1                          2   
3          1          2      3                          1   
4          0          0      3                          1   

   phys_health_consequence  coworkers  supervisor  mental_health_interview  \
0                        1          2           2                        1   
1                        1          1           2                        1   
2                        1          0           1                        1   
3                        1          1           1                        1   
4                        1          1           1                        1   

   phys_health_interview  mental_vs_physical  obs_consequence  
0                      0                   2                1  
1                      0                   1                1  
2                      0                   1                0  
3                      0                   1                0  
4                      1                   0                0

### Creating x and y

In [94]:
# Target/Dependent variable is Treatment
x = df.loc[:, df.columns != 'treatment']
y = df.loc[:,'treatment']

In [95]:
print(x.shape)
print(y.shape)

(1048, 22)
(1048,)


### Scaling the data

In [96]:
scaler = StandardScaler()
#X = scaler.fit_transform(X)
#sepatate function for fit and transform
scaler.fit(x)
x = scaler.transform(x)

In [97]:
scaler.fit(df_test)
df_test = scaler.transform(df_test)

### Splitting the data into train and test set

Splitting train set to train and validation set as we already have our test set

In [98]:
x_train,x_test,y_train,y_test = train_test_split(x,y,test_size=0.3,random_state = 10,stratify = y)

In [99]:
print("x_train:",x_train.shape)
print("x_test:",x_test.shape)
print("y_train:",y_train.shape)
print("y_test:",y_test.shape)

x_train: (733, 22)
x_test: (315, 22)
y_train: (733,)
y_test: (315,)


### Building Logistic Regression Model

In [100]:
#training the model
lr = LogisticRegression()
lr.fit(x_train,y_train)

LogisticRegression()

In [101]:
#Looking for coefficients value of output equation as logistic regression outputs a equation too
Coefficients = list(zip(df.columns[:-1],lr.coef_.ravel())) #ravel() is used to flatten the array
print(Coefficients)
print(lr.intercept_)

[('Age', 0.18691004938388583), ('Gender', -0.22613078484951563), ('self_employed', -0.09299077687778105), ('family_history', 0.5666618556491801), ('treatment', 1.44995533851218), ('work_interfere', -0.1438376979965696), ('no_employees', -0.004869578411435209), ('remote_work', -0.1355992613036288), ('tech_company', 0.2571979610273078), ('benefits', 0.3173676851814538), ('care_options', -0.03904837938354557), ('wellness_program', -0.17608887706822662), ('seek_help', 0.31007714147465604), ('anonymity', -0.1144205454463344), ('leave', -0.021955331408659785), ('mental_health_consequence', 0.08248361613282175), ('phys_health_consequence', 0.24655653329045507), ('coworkers', -0.1516408777666067), ('supervisor', 0.06556151156797575), ('mental_health_interview', 0.23514030907842515), ('phys_health_interview', 0.08866204635728568), ('mental_vs_physical', 0.20286508907732181)]
[-0.14281957]


In [102]:
#outputting Probability matrix
prob = lr.predict_proba(x_test)
prob

array([[0.87282332, 0.12717668],
       [0.62618737, 0.37381263],
       [0.27380911, 0.72619089],
       [0.88789017, 0.11210983],
       [0.24226045, 0.75773955],
       [0.52363937, 0.47636063],
       [0.59353627, 0.40646373],
       [0.91453153, 0.08546847],
       [0.29152476, 0.70847524],
       [0.98600704, 0.01399296],
       [0.67205002, 0.32794998],
       [0.46011936, 0.53988064],
       [0.03351478, 0.96648522],
       [0.07060646, 0.92939354],
       [0.15139134, 0.84860866],
       [0.19851706, 0.80148294],
       [0.07828237, 0.92171763],
       [0.81882519, 0.18117481],
       [0.95251728, 0.04748272],
       [0.92393458, 0.07606542],
       [0.85602178, 0.14397822],
       [0.94362628, 0.05637372],
       [0.71631142, 0.28368858],
       [0.1217623 , 0.8782377 ],
       [0.30325848, 0.69674152],
       [0.01162083, 0.98837917],
       [0.10127109, 0.89872891],
       [0.61253417, 0.38746583],
       [0.62622571, 0.37377429],
       [0.0484211 , 0.9515789 ],
       [0.

In [103]:
#Predicting values
y_pred = lr.predict(x_test)

### Evaluation of model

In [104]:
cfm = confusion_matrix(y_test,y_pred)
print("Confusion Matrix: ")
print(cfm)

acc = accuracy_score(y_test,y_pred)
print("accuracy score",acc)

print('classification report : ')
print(classification_report(y_test,y_pred))

Confusion Matrix: 
[[134  27]
 [ 23 131]]
accuracy score 0.8412698412698413
classification report : 
              precision    recall  f1-score   support

           0       0.85      0.83      0.84       161
           1       0.83      0.85      0.84       154

    accuracy                           0.84       315
   macro avg       0.84      0.84      0.84       315
weighted avg       0.84      0.84      0.84       315



### Fine Tuning the model

#### 1.) Adjusting Threshold

Using Ridge Regression:

In [105]:
#to loop through various values of threshold and check which one gives least error(wrong predictions)
for i in np.arange(0.4,0.61,0.01):
    pred = np.where(prob[:,1]>i,1,0)
    cfm = confusion_matrix(y_test,pred)
    err = cfm[1,0]+cfm[0,1]
    print("Error for threshold,",i," : ",err, "and Type II error :",cfm[1,0],"and Type I error :",cfm[0,1])

Error for threshold, 0.4  :  46 and Type II error : 14 and Type I error : 32
Error for threshold, 0.41000000000000003  :  46 and Type II error : 15 and Type I error : 31
Error for threshold, 0.42000000000000004  :  47 and Type II error : 16 and Type I error : 31
Error for threshold, 0.43000000000000005  :  46 and Type II error : 16 and Type I error : 30
Error for threshold, 0.44000000000000006  :  45 and Type II error : 16 and Type I error : 29
Error for threshold, 0.45000000000000007  :  45 and Type II error : 17 and Type I error : 28
Error for threshold, 0.4600000000000001  :  45 and Type II error : 17 and Type I error : 28
Error for threshold, 0.4700000000000001  :  45 and Type II error : 17 and Type I error : 28
Error for threshold, 0.4800000000000001  :  47 and Type II error : 20 and Type I error : 27
Error for threshold, 0.4900000000000001  :  50 and Type II error : 23 and Type I error : 27
Error for threshold, 0.5000000000000001  :  50 and Type II error : 23 and Type I error : 2

We can see we are getting lesser Type 2 error at threshold 0.44 so we'll try prdeicting the value using that threshold and generate evaluation metrics

In [106]:
#prob is the probability matrix that we genearted above
y_pred_t = [1 if i>0.44 else 0 for i in prob[:,1]]

In [107]:
cfm = confusion_matrix(y_test,y_pred_t)
print("Confusion Matrix: ")
print(cfm)

acc = accuracy_score(y_test,y_pred_t)
print("accuracy score",acc)

print('classification report : ')
print(classification_report(y_test,y_pred_t))

Confusion Matrix: 
[[132  29]
 [ 16 138]]
accuracy score 0.8571428571428571
classification report : 
              precision    recall  f1-score   support

           0       0.89      0.82      0.85       161
           1       0.83      0.90      0.86       154

    accuracy                           0.86       315
   macro avg       0.86      0.86      0.86       315
weighted avg       0.86      0.86      0.86       315



We can conclude following from above evaluation:

    Recall for class 0 has been decresed by 1 unit but f1-score is  increased by 1 unit
    
    Recall for class 1 has been increased by 6 units and f1-score is  increased by 2 units
    
    Accuracy score has improved a little bit from 0.838 to 0.857

Therefore, we'll use the logistic regression model having threshold as 0.44

In [108]:
log_adj = LogisticRegression()
lr.fit(x_train,y_train)

LogisticRegression()

### SGD Classification

In [109]:
sgd = SGDClassifier(loss = 'log', random_state = 10, 
                    learning_rate ='constant', eta0 = 0.00001,
                   max_iter= 1000,shuffle = True,
                   early_stopping = True,n_iter_no_change = 5)
sgd.fit(x_train,y_train)
y_sgd = sgd.predict(x_test)

In [110]:
cfm = confusion_matrix(y_test,y_sgd)
print("Confusion Matrix: ")
print(cfm)

acc = accuracy_score(y_test,y_sgd)
print("accuracy score",acc)

print('classification report : ')
print(classification_report(y_test,y_sgd))

Confusion Matrix: 
[[130  31]
 [ 32 122]]
accuracy score 0.8
classification report : 
              precision    recall  f1-score   support

           0       0.80      0.81      0.80       161
           1       0.80      0.79      0.79       154

    accuracy                           0.80       315
   macro avg       0.80      0.80      0.80       315
weighted avg       0.80      0.80      0.80       315



As we can see, accuracy score and f1-score for both the classes are lower than sklearn model so we'll not use this

# <font size = 5 color='Black'><b>Therefore, we can conclude that the best model is the one built using Logistic  Regression<b> 

Predicting values for actual test data

In [113]:
#lr = LogisticRegression()
#lr.fit(x_train,y_train)
prob_test = lr.predict_proba(df_test)
prob_test

array([[0.0443497 , 0.9556503 ],
       [0.06181987, 0.93818013],
       [0.21135559, 0.78864441],
       [0.61472557, 0.38527443],
       [0.91092196, 0.08907804],
       [0.61114726, 0.38885274],
       [0.03880332, 0.96119668],
       [0.96947153, 0.03052847],
       [0.96782449, 0.03217551],
       [0.48805494, 0.51194506],
       [0.19981796, 0.80018204],
       [0.85914125, 0.14085875],
       [0.35227147, 0.64772853],
       [0.92234828, 0.07765172],
       [0.92546411, 0.07453589],
       [0.22091041, 0.77908959],
       [0.46188839, 0.53811161],
       [0.9354786 , 0.0645214 ],
       [0.89279781, 0.10720219],
       [0.74189977, 0.25810023],
       [0.98300291, 0.01699709],
       [0.33201521, 0.66798479],
       [0.36866554, 0.63133446],
       [0.31680482, 0.68319518],
       [0.96517363, 0.03482637],
       [0.28092569, 0.71907431],
       [0.04643223, 0.95356777],
       [0.11692543, 0.88307457],
       [0.17727679, 0.82272321],
       [0.06625017, 0.93374983],
       [0.

In [114]:
pred_test =[1 if i>0.44 else 0 for i in prob_test[:,1]]

In [115]:
pred_test

[1,
 1,
 1,
 0,
 0,
 0,
 1,
 0,
 0,
 1,
 1,
 0,
 1,
 0,
 0,
 1,
 1,
 0,
 0,
 0,
 0,
 1,
 1,
 1,
 0,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 0,
 1,
 0,
 1,
 1,
 0,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 0,
 0,
 0,
 1,
 1,
 1,
 0,
 0,
 1,
 1,
 0,
 0,
 1,
 0,
 0,
 1,
 0,
 1,
 0,
 1,
 1,
 0,
 1,
 1,
 0,
 1,
 0,
 1,
 1,
 1,
 1,
 0,
 1,
 0,
 1,
 1,
 1,
 1,
 1,
 0,
 1,
 1,
 1,
 1,
 1,
 0,
 0,
 1,
 1,
 1,
 1,
 0,
 1,
 0,
 1,
 1,
 0,
 0,
 0,
 0,
 1,
 0,
 1,
 0,
 0,
 1,
 1,
 1,
 1,
 0,
 0,
 1,
 0,
 0,
 1,
 1,
 0,
 0,
 1,
 0,
 1,
 1,
 1,
 1,
 0,
 1,
 0,
 0,
 1,
 1,
 0,
 1,
 1,
 1,
 0,
 0,
 0,
 1,
 1,
 0,
 0,
 0,
 0,
 1,
 0,
 0,
 0,
 0,
 1,
 1,
 0,
 1,
 0,
 0,
 1,
 1,
 1,
 0,
 0,
 0,
 0,
 1,
 1,
 1,
 1,
 0,
 1,
 1,
 0,
 1,
 1,
 0,
 1,
 1,
 1,
 0,
 0,
 0,
 0,
 1,
 1,
 1,
 1,
 0,
 0,
 1,
 0,
 1,
 0,
 0,
 1,
 1,
 0,
 1,
 1,
 0,
 0,
 1,
 1,
 0,
 1]

In [125]:
output=pd.read_csv(r"C:/Users/odhonde/OneDrive - DXC Production/Documents/Imeticas Learning/Hackthon_13_05_2023/employee_test.csv", delimiter = ',')
output.columns=['S.No', 'Timestamp', 'Age', 'Gender', 'Country', 'state',
       'self_employed', 'family_history', 'work_interfere', 'no_employees',
       'remote_work', 'tech_company', 'benefits', 'care_options',
       'wellness_program', 'seek_help', 'anonymity', 'leave',
       'mental_health_consequence', 'phys_health_consequence', 'coworkers',
       'supervisor', 'mental_health_interview', 'phys_health_interview',
       'mental_vs_physical', 'obs_consequence', 'comments']
#output = pd.DataFrame()
output['treatment'] = pred_test
output['treatment'] = output['treatment'].map({1:"Yes",0:"No"})

In [126]:
output

S.No         Timestamp  Age  Gender         Country state self_employed  \
0       1   8/29/2014 11:32   39    Male  United Kingdom   NaN           Yes   
1       2   8/29/2014 11:32   26  female   United States    WA            No   
2       3   8/29/2014 11:33   23  Female   United States    IL            No   
3       4   8/29/2014 11:34   35    Male     Switzerland   NaN            No   
4       5   8/29/2014 11:36   36    Male   United States    FL            No   
..    ...               ...  ...     ...             ...   ...           ...   
205   206   9/12/2015 11:17   26    male  United Kingdom   NaN            No   
206   207    9/26/2015 1:07   32    Male   United States    IL            No   
207   208   11/7/2015 12:36   34    male   United States    CA            No   
208   209  11/30/2015 21:25   46       f   United States    NC            No   
209   210    2/1/2016 23:04   25    Male   United States    IL            No   

    family_history work_interfere    no_employees remote_work tech_company  \
0              Yes      Sometimes           5-Jan         Yes          Yes   
1              Yes      Sometimes  More than 1000          No          Yes   
2              Yes      Sometimes          26-100          No           No   
3              Yes          Often  More than 1000          No          Yes   
4               No          Never           5-Jan         Yes          Yes   
..             ...            ...             ...         ...          ...   
205             No            NaN          26-100          No          Yes   
206            Yes          Often          26-100         Yes          Yes   
207            Yes      Sometimes  More than 1000          No          Yes   
208             No            NaN         100-500         Yes          Yes   
209            Yes      Sometimes          26-100          No           No   

       benefits care_options wellness_program   seek_help   anonymity  \
0            No          Yes              Yes         Yes         Yes   
1           Yes          Yes               No          No  Don't know   
2            No           No               No          No  Don't know   
3    Don't know     Not sure               No          No         Yes   
4    Don't know     Not sure       Don't know  Don't know  Don't know   
..          ...          ...              ...         ...         ...   
205          No           No               No          No  Don't know   
206         Yes          Yes               No          No         Yes   
207         Yes          Yes               No          No  Don't know   
208          No          Yes               No          No  Don't know   
209         Yes          Yes               No          No         Yes   

                  leave mental_health_consequence phys_health_consequence  \
0    Somewhat difficult                        No                      No   
1            Don't know                        No                      No   
2    Somewhat difficult                       Yes                      No   
3             Very easy                        No                      No   
4             Very easy                        No                      No   
..                  ...                       ...                     ...   
205       Somewhat easy                        No                      No   
206  Somewhat difficult                        No                      No   
207  Somewhat difficult                       Yes                     Yes   
208          Don't know                       Yes                      No   
209          Don't know                     Maybe                      No   

        coworkers    supervisor mental_health_interview phys_health_interview  \
0             Yes           Yes                      No                 Maybe   
1    Some of them           Yes                      No                 Maybe   
2              No  Some of them                      No       

In [128]:
output.to_excel("OutPutLR.xlsx",header=True,index=False)